### PR TITLE
feat: add support for aggregates and toxicity classification

### PR DESCRIPTION
**Pull request by jarulraj**
1. Based on https://github.com/georgia-tech-db/eva/pull/519
2. Brings back pip package testing in CI
3. Reduces verbosity of YOLO model
4. Update links in readme to point to the `stable` version in read-the-docs
5. Add support for toxicity detection based on https://github.com/georgia-tech-db/eva/pull/516
6. Add support for querying based on video timestamps (based on https://github.com/georgia-tech-db/eva/pull/520)

### CODE REVIEW

File: eva/parser/lark_visitor/_functions.py [hunk 1]
```diff
@@ -125,22 +136,24 @@ def get_aggregate_function_type(self, agg_func_name):
...
+
         for child in tree.children:
             if isinstance(child, Tree):
                 if child.data == "function_arg":
                     agg_func_arg = self.visit(child)
+                elif child.data == "aggregate_function_name":
+                    agg_func_name = self.visit(child).value
+            elif isinstance(child, Token):
+                token = child.value
+                # Support for COUNT(*)
```
Comment (gaurav274) [inline eva/parser/lark_visitor/_functions.py:151]: I don't understand this logic. Are we hardcoding `*` to `id` in the parser? If yes, I guess even though it is hacky, it saves us from handling this corner case in the binder. We could change it to `IDENTIFIER_COLUMN`, which is supposed to be a unique row id in all the tables.

File: eva/models/storage/batch.py [hunk 2]
```diff
@@ -378,7 +380,14 @@ def aggregate(self, method: str) -> None:
         Arguments:
             method: string with one of the five above options
         """
-        self._frames = self._frames.agg([method])
+        # Aggregate ndarray
+        if isinstance(self._frames.iat[0, 0], np.ndarray):
```
Comment (gaurav274) [inline eva/models/storage/batch.py:384]: Is it aggregating each row of the array? If yes, I suspect that will break the execution logic.

File: test/integration_tests/test_select_executor.py [hunk 3]
```diff
@@ -247,6 +247,21 @@ def test_select_and_limit(self):
...
         self.assertEqual(actual_batch, expected_batch[0])
 
+    def test_select_and_aggregate(self):
+        simple_aggregate_query = "SELECT COUNT(*), AVG(id) FROM MyVideo;"
+        actual_batch = execute_query_fetch_all(simple_aggregate_query)
+
+        self.assertEqual(actual_batch.frames.iat[0, 0], 10)
+        self.assertEqual(actual_batch.frames.iat[0, 1], 4.5)
+
+        complex_aggregate_query = """SELECT SUM(id), COUNT(label)
```
Comment (gaurav274) [inline test/integration_tests/test_select_executor.py:257]: We should add a test case with aggregate on ndarray column.

File: eva/parser/lark_visitor/_functions.py (hunk 1 above)
Comment (jarulraj) [reply eva/parser/lark_visitor/_functions.py:151]: Yes. This query currently works. I was also worried if "id" will be always present. How should I change it to `IDENTIFIER_COLUMN`?

File: eva/models/storage/batch.py (hunk 2 above)
Comment (jarulraj) [reply eva/models/storage/batch.py:384]: Yes, how will it break the execution logic?

The NDARRAY case is for object detection array etc. The normal case is the one that existed earlier -- `self._frames = self._frames.agg([method])`

File: test/integration_tests/test_select_executor.py (hunk 3 above)
Comment (jarulraj) [reply test/integration_tests/test_select_executor.py:257]: A weird part is that when the query operates on an ndarray column, it does not reduce it to a single row. It actually keeps as many rows around as the number of the rows in the input column. Does that even make sense?

Comment (jarulraj) [reply test/integration_tests/test_select_executor.py:257]: Should we support the `NDARRAY` case?

File: eva/parser/lark_visitor/_functions.py (hunk 1 above)
Comment (jarulraj) [reply eva/parser/lark_visitor/_functions.py:151]: The unit tests do not create tables using `IDENTIFIER_COLUMN` -- so the test case fails.

File: test/integration_tests/test_select_executor.py (hunk 3 above)
Comment (jarulraj) [reply test/integration_tests/test_select_executor.py]: I have reverted back the `NDARRAY` case as it does not make sense.

File: eva/models/storage/batch.py (hunk 2 above)
Comment (jarulraj) [reply eva/models/storage/batch.py]: I have reverted back the `NDARRAY` case as it does not make sense.

**Comment by jarulraj** [comment]
I just added a `ToxicityClassifier` UDF -- that runs on top of OCR labels.

File: eva/parser/lark_visitor/_functions.py (hunk 1 above)
Comment (gaurav274) [reply eva/parser/lark_visitor/_functions.py:151]: Yeah, I verified we don't support projecting `IDENTIFIER_COLUMN`, which causes the binder to fail.
`id` won't work for images or other tables. Ideally, the binder should take care of it. 
An `if` condition  [here](https://github.com/georgia-tech-db/eva/blob/04368ef72a04973b5ad3bfdf98b34e957440539e/eva/binder/statement_binder.py#L103) should fix it.

Comment (gaurav274) [reply eva/parser/lark_visitor/_functions.py:151]: Reference: https://github.com/georgia-tech-db/eva/blob/04368ef72a04973b5ad3bfdf98b34e957440539e/eva/binder/binder_utils.py#L77

File: eva/models/storage/batch.py (hunk 2 above)
Comment (gaurav274) [reply eva/models/storage/batch.py]: Aggregate on `ndarray` and primitive column will result in different row counts. 

We have a different set of aggregate operators to apply row-wise aggregates like`Array_Count`

File: test/integration_tests/test_select_executor.py (hunk 3 above)
Comment (gaurav274) [reply test/integration_tests/test_select_executor.py]: Do we raise an error if the query tries to aggregate on an array column? We should add a test case to verify it. Thanks!

File: eva/expression/aggregation_expression.py [hunk 4]
```diff
@@ -55,9 +55,14 @@ def evaluate(self, *args, **kwargs):
         elif self.etype == ExpressionType.AGGREGATION_MAX:
```
Comment (gaurav274) [inline eva/expression/aggregation_expression.py:55]: We miss an `else` condition to raise an error. Right now, we silently ignore it and return the origin batch.

File: eva/readers/opencv_reader.py [hunk 5]
```diff
@@ -59,7 +59,11 @@ def _read(self) -> Iterator[Dict]:
                 _, frame = video.read()
                 frame_id = begin
                 while frame is not None and frame_id <= end:
-                    yield {"id": frame_id, "data": frame}
+                    yield {
+                        "id": frame_id,
+                        "data": frame,
+                        "seconds": frame_id // video.get(cv2.CAP_PROP_FPS),
```
Comment (gaurav274) [inline eva/readers/opencv_reader.py:65]: Are we planning to add a `seconds` column? If yes, we need to register it in the catalog, or the rest of the system won't be aware of it. Add it https://github.com/georgia-tech-db/eva/blob/04368ef72a04973b5ad3bfdf98b34e957440539e/eva/catalog/catalog_utils.py#L27

Btw we should not hardcode these strings in the reader, but rather get them from the catalog object. Something to do in the future. I'll take note of it.

File: eva/udfs/ndarray/timestamp.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# coding=utf-8
```
Comment (gaurav274) [inline eva/udfs/ndarray/timestamp.py:1]: What is this for?

File: eva/udfs/toxicity_classifier.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        # reconstruct dimension of the input
+        outcome = pd.DataFrame()
+        dataframe_size = text_dataframe.size
+
+        for i in range(0, dataframe_size):
+            text = text_dataframe.iat[i, 0]
+            single_result = self.model.predict(text)
+            toxicity_score = single_result["toxicity"][0]
+            if toxicity_score >= self.threshold:
+                outcome = outcome.append({"labels": "toxic"}, ignore_index=True)
```
Comment (gaurav274) [inline eva/udfs/toxicity_classifier.py:75]: I changed it to use `list` for the append operation. `DataFrame ` throws a lot of warnings. You can refer to other udfs.

File: eva/udfs/ndarray/timestamp.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            second           timestamp
+        0   int           0   string
+        1   int           1   string
+        """
+
+        # Sanity check
+        if len(inp.columns) != 1:
+            raise ValueError("input must only contain one column (seconds)")
+
+        seconds = pd.DataFrame(inp[inp.columns[0]])
```
Comment (gaurav274) [inline eva/udfs/ndarray/timestamp.py:42]: Isn't it `no-op`?

File: test/catalog/test_catalog_manager.py [hunk 9]
```diff
@@ -76,6 +76,7 @@ def test_create_multimedia_table_catalog_entry(self, mock):
             ColumnDefinition(
                 "data", ColumnType.NDARRAY, NdArrayType.UINT8, [None, None, None]
```
Comment (gaurav274) [inline test/catalog/test_catalog_manager.py:77]: `array_dimension` should be a tuple.

**Comment by gaurav274** [comment]
> I just added a `ToxicityClassifier` UDF -- that runs on top of OCR labels.

Thanks! This will be a fun example to showcase 💯

File: eva/readers/opencv_reader.py (hunk 5 above)
Comment (jarulraj) [reply eva/readers/opencv_reader.py:65]: Yes, this is done.

File: eva/udfs/toxicity_classifier.py (hunk 7 above)
Comment (jarulraj) [reply eva/udfs/toxicity_classifier.py:75]: Did you push your change?

File: eva/udfs/ndarray/timestamp.py (hunk 8 above)
Comment (jarulraj) [reply eva/udfs/ndarray/timestamp.py:42]: This is a timestamp UDF.

Comment (jarulraj) [reply eva/udfs/ndarray/timestamp.py:42]: `SELECT id, seconds, Timestamp(seconds)
                          FROM MyVideo
                          WHERE Timestamp(seconds) <= "00:00:01";`

**Pull request merged**